### ISSUE TITLE
Spike: Is it possible to distribute HNSW Graphs?

### ISSUE THREAD

**Issue by etiennedi** [semi-technologies/weaviate#1164]
## Todos
* [x] build master<>slave setup where one gets all the writes, but other party gets replicated
  * This answers the Question: "How much less expensive is it to just "assemble" without having to figure out the positions
* [x] spread writes evenly across two (or n) graphs where all can read/write to achieve a masterless setup
* [x] simulate network by introducing random delays, etc.
  * We might be able to assume a guaranteed processing as this is something we can control via the replication sync. E.g. with a small clockskew we could sort by timestamp and make sure that the list is read from top to bottom.

## Goal
Assume the following setup:

1. Have n parallel HNSW graphs
2. On import choose on of the graphs in a round robin fashion
3. Determine the index position and connected edges
4. Publish the results to all other graphs
5. When receiving a message from the another graph merge into own graph 

### Questions:
1. When we now search the graphs, do they all provide same (or very similar) results?
2. Was the import performance significantly better than just building n independent graphs in parallel?

## Is this just sharding?

The above is explicitly different from sharding. The goal is not to distribute the dataset in chunks, but to end up with n identical "replicas" where the core work of importing (figuring out which nodes to connect) was only ever done by a single node. Additionally because the importing shard can be chosen at will, we would end up with masterless completely identical replicas.

This could still be combined with sharding, e.g. you could have 5 shards, each replicated three times, for a total of 15 shards owning 5 different parts of the data.

**Comment by etiennedi** [comment]
## Results:
* CPU time increases for a slave-only replica was about 10% over a single master replica
* Using 2 masterless identical replicas with random inserts has the same performance as master/slave on the same machine (this is expected, as one machine already maxes out a single replica)
* No measurable loss in quality when introducing random delays between 0 and 500ms to simulate slow networks
